### ISSUE TITLE
option to format other filetypes (eg. `toml`, `yaml`, `json`, `md`, etc.)

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ruff#10738]
it would be nice if ruff could format `toml`, `yaml` and `md` files.

i get that this may be considered out of scope for ruff, however in the nodejs ecosystem, the prettier formatter works on these filetypes in addition to js & ts files, which is useful because most projects contain files written in those languages.

**Comment by KotlinIsland** [comment]
you can use [dprint](https://dprint.dev/) with dprint-py: https://pypi.org/project/dprint-py/ 🚀🚀🚀

also, [prek 🚀](https://github.com/j178/prek) and [lefthook](https://github.com/evilmartians/lefthook) are really powerful next generation git-hook managers

**Comment by MichaReiser** [comment]
This is related to https://github.com/astral-sh/ruff/issues/9244

Yeah, that's something I want to do eventually and we have the infrastructure for it (`ruff_formatter` is entirely language agnostic). But there are some things we have to figure out (what parsers to use or implement our own? How to integrate it into existing commands) and implementing additional language is a large commitment

**Comment by KotlinIsland** [comment]
I'm serious about [dprint](https://dprint.dev/), I would imagine it would be a massive amount of work, when dprint is doing exactly what you are setting out to achieve.

**Comment by MichaReiser** [comment]
Integrating dprint as an optional formatter (that we call into when the project has installed it) is an option. But I'm unsure if that's the experience users seek. We need to do some designing around this and decide if we want to build out support for new languages ourselves or delegate to other formatters (even Prettier)

**Comment by KotlinIsland** [comment]
Ruffs formatter is already integrated into dprint https://github.com/dprint/dprint-plugin-ruff 🚀

**Comment by DetachHead** [comment]
dprint also doesn't have a yaml formatter

**Comment by Avasam** [comment]
> dprint also doesn't have a yaml formatter

I mean *technically* it has prettier as a process plugin. Which you can configure to run only on yaml.
As much as I dislike prettier for JS/TS, it does just fine for markup/static languages.

There's also pre-commit's https://github.com/macisamuele/language-formatters-pre-commit-hooks which satisfies all my toml, yaml, ini (haven't tried the others) needs.

dprint also still doesn't work through pre-commit's free version if you have an online shared config.

Not saying this to discourage the OP's request. I too would like it Ruff handled common python config format (.ini for historical reasons with flake and mypy, toml obviously from pyproject.toml). Just offering some options in the short term.

**Comment by inoa-jboliveira** [comment]
Again I failed my github search and here I am with a duplicate. 

Addressing the post 1st comment, toml file for me is almost nobrainer. It is essential part of python ecosystem including heavy usage in uv and ruff as configuration. For the other you have many different tools in pre-commit, prettier and so on. Also JSON is such an universal format that you can build your formatter in a couple minutes or ask chatgpt for one.

I would pick a format for toml file that is really just a cleanup with not many opinions (remove blank lines, extra or missing spaces, lines too long). For example it should not try to convert an object into a block or vice versa. Nor it should try to reorganize blocks or sort arrays. 

If any of those features ever become needed they should be their own separated issue.

**Comment by KotlinIsland** [comment]
> I mean _technically_ it has prettier as a process plugin

they have a yaml plugin now 🚀
https://dprint.dev/plugins/pretty_yaml/

**Comment by Avasam** [comment]
> > I mean _technically_ it has prettier as a process plugin
> 
> they have a yaml plugin now 🚀 [dprint.dev/plugins/pretty_yaml](https://dprint.dev/plugins/pretty_yaml/)

You beat me to it updating my message :) (I've updated it anyway)

**Comment by ashrub-holvi** [comment]
Yep, looks like dprint is a nice tool, but all pre-commit options are not well supported, see https://github.com/dprint/dprint/issues/442, perhaps https://github.com/dprint/dprint/issues/442#issuecomment-3166402560 is supported

**Comment by DetachHead** [comment]
imo this feature is no longer necessary now that [dprint can now be installed via pypi](https://pypi.org/project/dprint-py/), making it a viable solution for python projects.

**Comment by MichaReiser** [comment]
[Prettier's yaml implementation](https://github.com/prettier/prettier/blob/main/src/language-yaml/index.js) doesn't look too complicated and should be relatively straightforward to port, given that we use a very similar formatter infrastructure. 

The main open questions are:

* What Yaml parser to use (and how much this impacts ruff's binary size)
* How to configure yaml support
* How to make the formatter pick up yaml files in addition to python files

**Comment by ashrub-holvi** [comment]
For YAML author of pre-commit [saying](https://github.com/pre-commit/pre-commit-hooks/issues/640#issuecomment-897201069) it's not that actually easy to do it properly

**Comment by KotlinIsland** [comment]
> For YAML author of pre-commit [saying](https://github.com/pre-commit/pre-commit-hooks/issues/640#issuecomment-897201069) it's not that actually easy to do it properly

dprint has yaml support https://dprint.dev/plugins/pretty_yaml/

**Comment by kasir-barati** [comment]
Landed here, and JFYI:

```bash
uv run ruff format **/*.md --preview
```

But the results ain't satisfactory at all. E.g. if you have this table in your markdown it won't align table columns vertically:

```md
| Col1 | Col2 |
| - | - |
| value 1 | value 2 |
```

And it won't remove unnecessary new lines (you can have 10 new lines and as you know when the markdown is rendered all those new lines will only amount to a single new line).

```bash
(jager-python) ➜  jager-python git:(main) ✗ uv run ruff --version
ruff 0.15.5
```

**Comment by ntBre** [comment]
@kasir-barati The new Markdown formatting feature is only for formatting Python code blocks within Markdown files, not for formatting the Markdown itself. See https://docs.astral.sh/ruff/formatter/#markdown-code-formatting

**Comment by epage** [comment]
I've been experimenting with a `Cargo.toml` formatter which might be relevant for `pyproject.toml` formatting
- project: https://github.com/crate-ci/cargo-cargofmt/ (readme also has a bunch of relevant links)
- example rule: https://github.com/crate-ci/cargo-cargofmt/blob/main/src/formatting/trailing_spaces.rs
